store/index/checkpoint.go

```diff
@@ -0,0 +1,348 @@
+package index
+
+import (
+	"bufio"
+	"context"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"os"
+	"strconv"
+	"time"
+
+	logging "github.com/ipfs/go-log/v2"
+	"github.com/ipld/go-storethehash/store/types"
+)
+
+var log = logging.Logger("storethehash/index")
+
+const (
+	// checkpointInterval is how often to advance the checkpoint.
+	checkpointInterval = 2 * time.Minute
+	// checkpointRuntime is the maximum time to run checkpoint advancement if
+	// it does not find a current entry to stop on.
+	checkpointRuntime = time.Minute
+	// compactPercentThreshold is percent of the index that must be eliminated
+	// by the checkpoint for compaction to take place.
+	compactPercentThreshold = 33
```

Comment:
these probably need to be changeable at least as flags if not during runtime

Reply:
I can probably get rid of `checkpointRuntime` (see next response). 

If we can come up with a best threshold to trigger compaction, then maybe `compactPercentThreshold` can be removed.  So far